crates/matrix-sdk/tests/integration/encryption.rs

```diff
@@ -0,0 +1,237 @@
+                    "device_keys": device_keys
+                }))
+            })
+            .mount(&server)
+            .await;
+
+        Mock::given(method("POST"))
+            .and(path("/_matrix/client/unstable/keys/device_signing/upload"))
+            .respond_with(move |req: &Request| {
+                // Accept all cross-signing setups by default.
+                #[derive(Debug, serde::Deserialize)]
+                struct Parameters {
+                    master_key: Option<Raw<CrossSigningKey>>,
+                    self_signing_key: Option<Raw<CrossSigningKey>>,
+                    user_signing_key: Option<Raw<CrossSigningKey>>,
+                }
+
+                let params: Parameters = req.body_json().unwrap();
+                assert!(params.master_key.is_some());
+                assert!(params.self_signing_key.is_some());
+                assert!(params.user_signing_key.is_some());
+
+                // We don't do anything particular with the keys, at this point.
+
+                ResponseTemplate::new(200).set_body_json(json!({}))
+            })
+            .mount(&server)
+            .await;
+
+        let known_devices_clone = known_devices.clone();
+        let device_keys_clone = device_keys.clone();
+        Mock::given(method("POST"))
+            .and(path("/_matrix/client/unstable/keys/signatures/upload"))
+            .respond_with(move |req: &Request| {
+                #[derive(Debug, serde::Deserialize)]
+                #[serde(transparent)]
+                struct Parameters(BTreeMap<OwnedUserId, SignedKeys>);
+
+                let params: Parameters = req.body_json().unwrap();
+
+                let mut device_keys = device_keys_clone.lock().unwrap();
+                let known_devices = known_devices_clone.lock().unwrap();
+                for (user_id, signed_keys) in params.0 {
+                    for (key_id, signature) in signed_keys.iter() {
+                        if known_devices.contains(key_id) {
+                            device_keys
+                                .entry(user_id.clone())
+                                .or_default()
+                                .insert(key_id.to_owned(), signature.to_owned());
+                        }
+                    }
+                }
+
+                ResponseTemplate::new(200).set_body_json(json!({
+                    "failures": {}
+                }))
+            })
+            .mount(&server)
+            .await;
+
+        Self { server, known_devices, device_keys }
+    }
+
+    fn add_known_device(&mut self, device_id: &DeviceId) {
+        self.known_devices.lock().unwrap().insert(device_id.to_string());
+    }
+}
+
+async fn bootstrap_cross_signing_and_verify_device_test_helper(client: &Client) {
+    client.encryption().bootstrap_cross_signing(None).await.unwrap();
+
+    let status = client.encryption().cross_signing_status().await.unwrap();
+    assert!(status.has_master);
```

Comment:
Perhaps you'll want this from the secret storage PR: https://github.com/matrix-org/matrix-rust-sdk/pull/2621/files#diff-5ef8b27df74094f85e75c840dff614573eb9bf72f03663622facfcfb9aaba68cR112-R117

Reply:
Actually, that's now part of https://github.com/matrix-org/matrix-rust-sdk/pull/2758/.